### ISSUE TITLE
Incorrect error "you are banned" in stale session

### ISSUE THREAD

**Issue by m4z** [SimpleMachines/SMF#7761]
(I'm on mobile, and for some reason the issue template doesn't show up.)

On SMorg, when keeping a tab around for a couple of hours (for example when reading a single page of a longer thread step by step on mobile) and then coming back and clicking on one of the "n people liked this" links, instead of the usual list, I receive a popup with the text "${myname}, you are banned from this forum!" (or similar).

A more helpful message would tell me that my session expired, or even do the reload with some JS magic, if possible. Reloading the page usually solves the problem, not even a re-login is necessary.

**Comment by jdarwood007** [comment]
This is because of how the popup javascript is handling errors.  What you are seeing is how it interrupts the error as being banned, rather than needing auth.

**Comment by m4z** [comment]
A more fitting error would be $txt['cannot_view_likes'].

**Comment by DiegoAndresCortes** [comment]
Would probably need to first add the text in here
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Sources/Load.php#L2565-L2589

The next issue is reworking this area
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Themes/default/scripts/script.js#L337
Since it only has that one status/state, it doesn't have distinct errors.

**Comment by jdarwood007** [comment]
Correct, as you can see in that code, when a 403 status occurs, it displays that message.  A non 403 needs to be sent.  Perhaps a 401 should be sent.  Maybe 401 with a hard coded text and 400 can have a custom text passed to it by sending back the text in a response header.

**Comment by DiegoAndresCortes** [comment]
Also the message should be different because they can see the likes, it's just the session that expired.
In fact I think it should be something more generic even. Some other features or mods could be using this overlay and could trigger the message it because the user lost permissions, not just because the session expired.

**Comment by DiegoAndresCortes** [comment]
Well I think the user doesn't need to know that information (permissions). So could be something like this:
`Your session may have expired, please refresh the page and try again.`

Edit: We already have a text string for this
$txt['session_verify_fail']

**Comment by jdarwood007** [comment]
Sounds good.  Want to implant this?  Shouldn't be too bad, add the js txt string and have it trigger for 401, ensure that for errors that are failing like this, that 401 is sent.  Maybe a future version can support sending a custom header to include custom text strings for error messages.

**Comment by DiegoAndresCortes** [comment]
It's not as simple at the moment, the likes are not sending the error, it's checkSession which doesn't receive an error status for fatal_lang_error, thus defaulting to 403.
Should checkSession have an additional parameter for the error status as well?

This is looking like it would be out of the scope for a patch.

**Comment by jdarwood007** [comment]
Well @sbulen would have to clarify.  I think then we could send a custom X-header with the text string.  fatal_lang_error can send it.  Then we parse for it and handle it as the message.

**Comment by sbulen** [comment]
Do we know what the actual error is?  Why is it failing at all?  Why can't it just return the info?

**Comment by DiegoAndresCortes** [comment]
Because there is no status for session defined here:
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Sources/Subs.php#L8161-L8171
So the thought is to add a 401 for session failure and include it up there, and here:
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Themes/default/scripts/script.js#L373-L380

And the problem with that, is that while you can send an custom status to the function:
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Sources/Errors.php#L194

Likes action is not sending a fatal_error, is just doing a session check.
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Sources/Likes.php#L147

But checkSession sends 403 as a default error code:
https://github.com/SimpleMachines/SMF/blob/78e16b554272cf84d42a4b3401656972022e469a/Sources/Security.php#L710

So apart from adding the new status there are two options:

1. Add a new parameter to checkSession so that it can pass it to fatal_lang_error.
2. Directly pass a 401 in fatal_lang_error when it's called from checkSession.

**Comment by sbulen** [comment]
Thanks for the info.

But let me rephrase my question: Is the session really expired?  He doesn't need to log back in, just refresh the page.

You are correct it would really help to provide a better message.   

But that session has not expired, so why is there an error at all?

**Comment by DiegoAndresCortes** [comment]
It did expire.
The incorrect message (banned) appears in those popups when the session expires.
In the example provided in the issue, either the session expired or user signed out in another tab.
The other scenario is that the user lost permissions for X feature.

**Comment by sbulen** [comment]
In the initial description above, he indicates no new logon was necessary (last sentence).  Only a refresh.  So I don't think the session expired.

**Comment by DiegoAndresCortes** [comment]
Ah I missed that part. I could only reproduce the other two scenarios consistently.

**Comment by sbulen** [comment]
If I open multiple tabs, and logout on one, and click a like on the other, I get a "There was a problem with your request", and refreshing the screen shows me I'm logged out.

That's acceptable, I think.  

The banned error is not...  (Unless you were actually banned...)

**Comment by DiegoAndresCortes** [comment]
It’s for the liked by list, this doesn’t happen when liking.
The issue is with popups/overlays.

**Comment by BrickOzp** [comment]
Could it be the PHP session that has expired, but the SMF session is still valid?

**Comment by jdarwood007** [comment]
I sent up a PR which helps this by sending a custom header.  Due to language not being loaded, I have to move the loading of the language earlier.  Hopefully another bug isn't created by this.

The SM.org issue I assume is because the session has expired, then your trying to use it.  The system rejects it as expected.  Because your session ended, the next call rebuilds it.  Then when checkSession comes around, it finds that your new session doesn't match the session data sent over.  This issue is separate from the SMF issue and most likely something with our site configuration if the keepalive isn't maintaining your session.

**Comment by m4z** [comment]
If it helps, I usually never log out of forums, but I switch between many devices and browsers over the course of a typical day and even use them simultaneously.

**Comment by jdarwood007** [comment]
Your devices here are not really relevant.  What you see is normal.  Your session expired on the server.  Normal browsing just builds it with your cookies to login again.  However since it was a request that we validate sessions on, it failed.   This would happen to normal pages we validate the session on as well.
If the tab/window is inactive, keepalive most likely is not sent (browser dependent to send keepalives).  Which means that the server can't be told to keep the session alive and going.

You can verify this on any page that requires session checking by modifying the session token.  It will produce the same error as if you had let the session expire.

The sequence seen here is:
1. Your request comes in from a stale browsing session.
2. Your session is expired, you can't continue, are no longer "logged" and any other session based data is lost.
3. New clean session information is built automatically.
4. Login is successful, using cookies sent in the request.
5. checkSession is called
6. checkSession finds the token is not valid, because the request sent the old session data.
7. checkSession causes fatal error and stops.
